query-planner-js/src/utilities/debug.ts

```diff
@@ -0,0 +1,143 @@
+// Simple debugging facility.
+
+import chalk from 'chalk';
+
+function indentString(indentLevel: number) : string {
+  let str = "";
+  for (let i = 0; i < indentLevel; i++) {
+    str += chalk.blackBright("⎸ ");
+  }
+  return str;
+}
+
+/**
+ * A facility to log messages for debugging.
+ *
+ * A `DebugLogger` maintains an indentation level that is used before logged messages. That indentation level gets
+ * increased by calling `group()` and decreased by `groupEnd()`. This is meant to help representing the nestedness
+ * of the code being debugged.
+ */
+export class DebugLogger {
```

Comment:
I'm okay with this approach of _rolling our own logger_  (also, I want to brag that this saying works really well when talking about _logs_ 🌲) for this particular case because the scope of the work doesn't entail exposing this to the user right now and this solves a concrete need as soon as possible.

My hope would be that, in the future we would be willing to consider an approach that exposes and allows our users to utilize the common logger facilities they already enjoy (defined on the `logger` property of the top-level `ApolloServer` constructor), including those that support JSON logging to a common log facility, etc. (In particular, I can imagine JSON logging being a really great format for capturing/representing the depth of these!)

That said, I do want to draw attention to the [standardized Console API](https://console.spec.whatwg.org/) though, which does support [automatic nesting/grouping of concerns / automatic indentation](https://developer.mozilla.org/en-US/docs/Web/API/Console#using_groups_in_the_console).  That API is also [implemented in Node.js itself](https://nodejs.org/api/console.html).  Of course, we're not just targeting Node.js, but the standardization of this `Console` API is somewhat prolific at this point.

There are _so many loggers_, but for the sake of capturing some options for the future:

* https://npm.im/debug
* https://npm.im/loglevel
* https://npm.im/bunyan
* https://npm.im/winston
* https://npm.im/logtown
* https://npm.im/log4js
* https://npm.im/pino
* https://npm.im/console-pro (this is the only one that I believe supports `groupEnd`!

Reply:
(Just to be clear though, I'm not blocking.)

Reply:
Thanks for the comment and link. I did answer some related things in my answer to David so I'll try to not repeat it.

I did knew about `console.group` and friends, and that's even why I use the same name. I still wanted to not have those logged by default (and passing closure to avoid having to pay the cost of string interpolation in at least some cases), so we needed a small facade. I happen to not use `console.group` within said facade, but it's just to get something a tad cleaner looking and is a minor implementation detail at this point.

I did miss [console-pro](https://npm.im/console-pro) in my search and it's the one that most ressemble what I'm doing here, and we can switch to that if we really prefer. That said, the `DebugLogger` here is really very little code and is a bit more tailored to our exact need, so I'm personally happy to stick to that for now.

> My hope would be that, in the future we would be willing to consider an approach that exposes and allows our users to utilize the common logger facilities they already enjoy (...)

Agreed, but to reiter my comment to David, I don't think what we do here is meant for users ever. Having debug logging that can be understood by users to help them optimize their queries (or at least understand why they get the plan they get) would definitively be great, this is just not what we do here :).